git_machete/cmd.py

```diff
@@ -2587,53 +1774,34 @@ def get_latest_checkout_timestamps(cli_ctxt: CommandLineContext) -> Dict[str, in
 
 # Complex routines/commands
 
-def is_merged_to(machete_client: MacheteClient, cli_ctxt: CommandLineContext, b: str, target: str) -> bool:
-    if is_ancestor_or_equal(cli_ctxt, b, target):
+def is_merged_to(machete_client: MacheteClient, b: str, target: str) -> bool:
+    if is_ancestor_or_equal(machete_client.git, b, target):
         # If branch is ancestor of or equal to the target, we need to distinguish between the
         # case of branch being "recently" created from the target and the case of
         # branch being fast-forward-merged to the target.
         # The applied heuristics is to check if the filtered reflog of the branch
         # (reflog stripped of trivial events like branch creation, reset etc.)
         # is non-empty.
         return bool(machete_client.filtered_reflog(b, prefix="refs/heads/"))
-    elif cli_ctxt.opt_no_detect_squash_merges:
+    elif machete_client.cli_opts.opt_no_detect_squash_merges:
         return False
     else:
         # In the default mode.
         # If there is a commit in target with an identical tree state to b,
         # then b may be squash or rebase merged into target.
-        return contains_equivalent_tree(cli_ctxt, b, target)
+        return contains_equivalent_tree(machete_client.git, b, target)
 
 
-def get_revision_repr(cli_ctxt: CommandLineContext, revision: str) -> str:
-    short_sha = short_commit_sha_by_revision(cli_ctxt, revision)
-    if is_full_sha(revision) or revision == short_sha:
+def get_revision_repr(git: GitContext, revision: str) -> str:
+    short_sha = git.short_commit_sha_by_revision(revision)
+    if git.is_full_sha(revision) or revision == short_sha:
         return f"commit {revision}"
     else:
-        return f"{revision} (commit {short_commit_sha_by_revision(cli_ctxt, revision)})"
-
-
-class StopTraversal(Exception):
-    def __init__(self) -> None:
-        pass
+        return f"{revision} (commit {git.short_commit_sha_by_revision(revision)})"
 
 
-def flush_caches() -> None:
-    global commit_sha_by_revision_cached, config_cached, counterparts_for_fetching_cached, initial_log_shas_cached
-    global local_branches_cached, reflogs_cached, remaining_log_shas_cached, remote_branches_cached
-    MacheteClient.branch_defs_by_sha_in_reflog = None
-    commit_sha_by_revision_cached = None
-    config_cached = None
-    counterparts_for_fetching_cached = None
-    initial_log_shas_cached = {}
-    local_branches_cached = None
-    reflogs_cached = None
-    remaining_log_shas_cached = {}
-    remote_branches_cached = None
-
-
-def pick_remote(cli_ctxt: CommandLineContext, b: str) -> None:
-    rems = remotes(cli_ctxt)
+def pick_remote(git: GitContext, b: str) -> None:
```

Comment:
Move into `MacheteClient` (this could be a private, used only by `traverse` under the hood)